### ISSUE TITLE
DeepSeek V4 缺少与 model.py 对齐的 RoPE/YaRN real cos/sin table 生成逻辑

### ISSUE THREAD

**Issue by wuzhf9** [hw-native-sys/pypto-lib#536]
## 背景

`model.py` 中的 DeepSeek attention 使用 `precompute_freqs_cis()` 生成 RoPE/YaRN 频率表：

- 先按 `rope_theta` / `compress_rope_theta`、`rope_factor`、`beta_fast`、`beta_slow`、`original_seq_len` 计算 YaRN 修正后的角频率。
- 再通过 `torch.polar(torch.ones_like(freqs), freqs)` 生成 complex `freqs_cis`。
- `apply_rotary_emb()` 通过 `torch.view_as_complex()`、complex multiply 和 `torch.view_as_real()` 对 interleaved even/odd rotary pairs 做 RoPE。
- `inverse=True` 时使用 `freqs_cis.conj()`，用于 attention output rope slice 的 de-rotation。

当前 `models/deepseek/v4` 路径下的实现不使用 complex tensor。各 PyPTO kernel 已经通过显式 real-valued `freqs_cos` / `freqs_sin` 输入完成 RoPE apply，但仓库里还缺少一个与 `model.py` 语义等价的 host-side RoPE/YaRN table 生成模块。目前多个 standalone fixture 使用类似 `torch.cos(torch.arange(...) * 1e-3)` / `torch.sin(torch.arange(...) * 1e-3)` 的 synthetic table，这不是 DeepSeek YaRN 频率表。

因此需要补齐的是：在 V4 路径上实现 `model.py::precompute_freqs_cis()` 等价的 real cos/sin table 生成逻辑，并统一各模块的 fixture/runtime 调用。

### 1. 公共 Q/KV 投影

`models/deepseek/v4/qkv_proj_rope.py`

- `materialize_rope_rows(...)`
  - 输入 full table：
    - `freqs_cos: [MAX_SEQ_LEN, ROPE_DIM], BF16`
    - `freqs_sin: [MAX_SEQ_LEN, ROPE_DIM], BF16`
    - `position_ids: [T_DYN], INT32`
  - 输出 token-local table：
    - `rope_cos_t: [T_DYN, ROPE_DIM], BF16`
    - `rope_sin_t: [T_DYN, ROPE_DIM], BF16`
  - 只做按 position 取行，不做三角函数计算。

- `qkv_proj_rope(...)`
  - 输入 token-local table：
    - `rope_cos: [T_DYN, ROPE_DIM], BF16`
    - `rope_sin: [T_DYN, ROPE_DIM], BF16`
  - 对 Q head rope slice 和 KV rope slice 做正向 RoPE。
  - 内部通过 `j >> 1` duplicate gather cos/sin，通过 `j ^ 1` swap 和 sign mask 实现 interleaved pair rotation。

### 2. Decode attention 编排模块

`decode_attention_swa.py`

- `compress_ratio = 0`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 内部按 `position_ids[t]` materialize：
  - `rope_cos_t/rope_sin_t: [T, ROPE_HEAD_DIM], BF16`
- `rope_cos_t/rope_sin_t` 同时传给：
  - `qkv_proj_rope(...)`：Q/KV 正向 RoPE
  - `decode_sparse_attn.sparse_attn(...)`：attention output rope slice 的 inverse RoPE

`decode_attention_csa.py`

- `compress_ratio = 4`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 内部派生：
  - `rope_cos_t/rope_sin_t: [T, ROPE_HEAD_DIM], BF16`
  - `step_cos/step_sin: [B, ROPE_HEAD_DIM // 2], FP32`
  - `cmp_cos/cmp_sin: [B, ROPE_HEAD_DIM // 2], FP32`
- `rope_cos_t/rope_sin_t` 给 `qkv_proj_rope(...)` 和 `decode_sparse_attn.sparse_attn(...)`。
- `step_cos/step_sin` 给 `decode_indexer.indexer(...)`。
- `cmp_cos/cmp_sin` 给 `decode_compressor_ratio4.compressor_ratio4(...)`。

`decode_attention_hca.py`

- `compress_ratio = 128`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 内部派生：
  - `rope_cos_t/rope_sin_t: [T, ROPE_HEAD_DIM], BF16`
  - `cmp_cos/cmp_sin: [B, ROPE_HEAD_DIM // 2], FP32`
- `rope_cos_t/rope_sin_t` 给 `qkv_proj_rope(...)` 和 `decode_sparse_attn.sparse_attn(...)`。
- `cmp_cos/cmp_sin` 给 `decode_compressor_ratio128.compressor_ratio128(...)`。

### 3. Decode 子模块

`decode_compressor_ratio4.py`

- 不接收 full table。
- 接收上层已 materialize 的 half table：
  - `cos: [B, ROPE_HEAD_DIM // 2], FP32`
  - `sin: [B, ROPE_HEAD_DIM // 2], FP32`
- 对 pooled KV rope slice 做正向 RoPE，写入 compressed KV cache。

`decode_compressor_ratio128.py`

- 与 ratio-4 decode compressor 类似：
  - `cos: [B_DYN, ROPE_HEAD_DIM // 2], FP32`
  - `sin: [B_DYN, ROPE_HEAD_DIM // 2], FP32`
- 用于 ratio-128 compressed KV 写入路径。

`decode_indexer.py`

- 接收 per-batch half table：
  - `cos: [B, ROPE_HEAD_DIM // 2], FP32`
  - `sin: [B, ROPE_HEAD_DIM // 2], FP32`
- 对 indexer query 的 rope slice 做正向 RoPE。
- 注意：如果 decode path 中 `S > 1`，当前 per-batch half table 设计可能会让同一 batch 内多个 token 复用同一行 cos/sin，需要和 `model.py` 的逐 token `freqs_cis[start_pos:start_pos+seqlen]` 语义单独核对。

`decode_indexer_compressor.py`

- 接收 per-batch half table：
  - `cos: [B, ROPE_HEAD_DIM // 2], FP32`
  - `sin: [B, ROPE_HEAD_DIM // 2], FP32`
- 对 inner compressed KV rope slice 做正向 RoPE。

`decode_sparse_attn.py`

- 接收 token-local table：
  - `freqs_cos/freqs_sin: [T, ROPE_DIM], BF16`
- 对 attention output rope slice 做 inverse RoPE。
- 复用同一组 cos/sin table，只在 apply 公式中使用 conjugate 对应的符号。

### 4. Prefill attention 编排模块

`prefill_attention_swa.py`

- `compress_ratio = 0`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 调用 `materialize_rope_rows(...)` 得到 token-local `rope_cos_t/rope_sin_t`。
- token-local table 传给 `qkv_proj_rope(...)` 和 `prefill_sparse_attn(...)`。

`prefill_attention_csa.py`

- `compress_ratio = 4`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 调用 `materialize_rope_rows(...)` 得到 token-local table。
- full table 继续传给：
  - `prefill_compressor_ratio4(...)`
  - `prefill_indexer(...)`
- 子模块内部按 compressed write position 取 half-width cos/sin。

`prefill_attention_hca.py`

- `compress_ratio = 128`
- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 调用 `materialize_rope_rows(...)` 得到 token-local table。
- full table 继续传给 `prefill_compressor_ratio128(...)`。

### 5. Prefill 子模块

`prefill_compressor_ratio4.py`

- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 对每个 compressed write token 计算：
  - `cmp_pos = write_pos + 1 - COMPRESS_RATIO`
- 读取 `freqs_cos/freqs_sin[cmp_pos, :ROPE_HEAD_DIM // 2]`，对 pooled KV rope slice 做正向 RoPE。

`prefill_compressor_ratio128.py`

- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_DIM], BF16`
- 同样通过 `cmp_pos = write_pos + 1 - COMPRESS_RATIO` 读取 half-width cos/sin 行，对 ratio-128 pooled KV rope slice 做正向 RoPE。

`prefill_indexer.py`

- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 主要是编排模块，把 full table 继续传给 `prefill_indexer_compressor(...)`。

`prefill_indexer_compressor.py`

- 接收 full table：
  - `freqs_cos/freqs_sin: [MAX_SEQ_LEN, ROPE_HEAD_DIM], BF16`
- 内部同样通过 `cmp_pos = write_pos + 1 - COMPRESS_RATIO` 读取 half-width cos/sin 行，对 inner compressed KV rope slice 做正向 RoPE。

`prefill_sparse_attn.py`

- 接收 token-local table：
  - `freqs_cos/freqs_sin: [T, ROPE_DIM], BF16`
- 对 attention output rope slice 做 inverse RoPE。

## 与 `model.py` 的公式对齐

`model.py::apply_rotary_emb()` 的 complex multiply 等价于：

```text
正向:
even' = even * cos - odd * sin
odd'  = even * sin + odd * cos

inverse:
even' = even * cos + odd * sin
odd'  = odd * cos - even * sin
```

当前 V4 路径中：

- `qkv_proj_rope.py` 的 Q/KV RoPE 与正向公式一致。
- decode/prefill compressor 的 KV rope 与正向公式一致，只要调用方提供的 cos/sin position 正确。
- `decode_sparse_attn.py` 和 `prefill_sparse_attn.py` 的 output rope slice 与 inverse 公式一致。
- 主要需要额外确认的是 decode indexer/inner indexer compressor 的 per-batch half table 设计，在 `S > 1` 或 compressed position 不等于 first token position 时是否和 `model.py` 完全等价。

## 实现方案

新增 host-side helper module：

```text
models/deepseek/v4/rope_tables.py
```

该模块不定义 `pl.jit` kernel。它只负责 host/runtime 侧 table 生成与 materialization。

建议 API：

```python
def rope_profile_for_compress_ratio(config, compress_ratio):
    """Return (base, original_seq_len) for this layer profile."""


def build_deepseek_v4_rope_tables(
    config,
    compress_ratio,
    *,
    max_seq_len=None,
    rope_dim=None,
    dtype=torch.bfloat16,
    device=None,
):
    """Return (freqs_cos, freqs_sin), both shaped [max_seq_len, rope_dim]."""


def build_layer_rope_tables(config, layer_id, *, dtype=torch.bfloat16, device=None):
    """Select compress_ratio from config.compress_ratios[layer_id]."""


def materialize_token_rope_tables(freqs_cos, freqs_sin, position_ids):
    """Return token-local (cos, sin) tables shaped [T, rope_dim]."""
```

### Profile 选择

与 `model.py` 保持一致：

```text
compress_ratio == 0:
  base = config.rope_theta
  original_seq_len = 0
  # 禁用 YaRN

compress_ratio != 0:
  base = config.compress_rope_theta
  original_seq_len = config.original_max_position_embeddings
  # 当 original_seq_len > 0 时启用 YaRN 修正
```

因此 full-model runtime 至少应缓存两类表：

- SWA table：`compress_ratio = 0`
- compressed-layer table：`compress_ratio = 4` / `128`，二者 RoPE profile 相同，只要 config 和 max_seq_len/rope_dim 相同即可复用

推荐 cache key：

```text
(config.name, max_seq_len, rope_dim, compress_ratio == 0 ? "swa" : "compressed", dtype, device)
```

### 频率生成算法

保留 `model.py` 的 YaRN inverse-frequency 修正逻辑，但输出 real cos/sin，不生成 complex tensor。

```text
inv_freq[i] = 1 / base ** (2 * i / dim)
i in [0, dim / 2)
```

当 `original_seq_len > 0`：

```text
correction_dim(rot) =
    dim * log(original_seq_len / (rot * 2*pi)) / (2 * log(base))

low  = floor(correction_dim(beta_fast))
high = ceil(correction_dim(beta_slow))
low/high clamp 到 [0, dim - 1]

ramp[i] = clamp((i - low) / (high - low), 0, 1)
smooth[i] = 1 - ramp[i]

inv_freq[i] =
    inv_freq[i] / factor * (1 - smooth[i]) + inv_freq[i] * smooth[i]
```

然后：

```text
angle[pos, i] = pos * inv_freq[i]
cos_half = cos(angle)
sin_half = sin(angle)
```

最终返回：

```text
freqs_cos = concat(cos_half, cos_half)  # [MAX_SEQ_LEN, ROPE_HEAD_DIM]
freqs_sin = concat(sin_half, sin_half)  # [MAX_SEQ_LEN, ROPE_HEAD_DIM]
```

注意：

- inverse frequencies、angles、cos/sin 计算使用 FP32。
- 只在 cos/sin 完成后 cast 到 BF16。
- 不在 V4 path 中返回 complex `freqs_cis`。
- 不新增 PyPTO device kernel 来生成整张 RoPE table；table 是只依赖 config 的常量，适合 host/runtime 侧预计算后作为普通 device input 传给 kernel。

### 调用点迁移

1. 替换 `models/deepseek/v4` 中 fixture 的 synthetic `cos(arange * 1e-3)` / `sin(arange * 1e-3)` 初始化。

2. 对接收 full table 的模块，直接使用：

```python
freqs_cos, freqs_sin = build_deepseek_v4_rope_tables(config, compress_ratio)
```

3. 对 token-local 接口，使用：

```python
full_cos, full_sin = build_deepseek_v4_rope_tables(config, compress_ratio)
token_cos, token_sin = materialize_token_rope_tables(full_cos, full_sin, position_ids)
```

4. 对 decode compressor/indexer standalone test 的 `[B, ROPE_HEAD_DIM // 2]` 输入，应从 full table 中按对应 absolute position 取：

```python
cos = full_cos.index_select(0, batch_positions)[:, :ROPE_HEAD_DIM // 2].float()
sin = full_sin.index_select(0, batch_positions)[:, :ROPE_HEAD_DIM // 2].float()
```

其中 `batch_positions` 需要根据模块语义分别是：

- decode compressor：compressed write position
- decode indexer query：query token position
- decode inner indexer compressor：inner compressed write position

### 1. Table 生成与 `model.py::precompute_freqs_cis()` 对齐

新增 focused unit test，分别覆盖：

- `compress_ratio = 0`
  - `base = rope_theta`
  - `original_seq_len = 0`
  - YaRN disabled
- `compress_ratio = 4`
  - `base = compress_rope_theta`
  - `original_seq_len = original_max_position_embeddings`
  - YaRN enabled when original length > 0
- `compress_ratio = 128`
  - 与 compressed profile 一致

断言：

```python
freqs_cis = model.precompute_freqs_cis(
    dim=rope_dim,
    seqlen=max_seq_len,
    original_seq_len=expected_original_seq_len,
    base=expected_base,
    factor=config.rope_factor,
    beta_fast=config.beta_fast,
    beta_slow=config.beta_slow,
)

freqs_cos, freqs_sin = build_deepseek_v4_rope_tables(...)

assert_allclose(freqs_cos[:, :rope_dim // 2].float(), freqs_cis.real.float())
assert_allclose(freqs_sin[:, :rope_dim // 2].float(), freqs_cis.imag.float())
assert_allclose(freqs_cos[:, rope_dim // 2:].float(), freqs_cos[:, :rope_dim // 2].float())
assert_allclose(freqs_sin[:, rope_dim // 2:].float(), freqs_sin[:, :rope_dim // 2].float())
```

### 2. 正向 RoPE apply 与 `model.py::apply_rotary_emb()` 对齐

构造随机 BF16 tensor：

- Q path shape example: `[B, S, H, ROPE_HEAD_DIM]`
- KV path shape example: `[B, S, ROPE_HEAD_DIM]`

用 `model.py::apply_rotary_emb(x, freqs_cis)` 得到 reference。

再用 real cos/sin 展开公式得到 V4 语义结果：

```python
x_pair = x.float().unflatten(-1, (-1, 2))
even = x_pair[..., 0]
odd = x_pair[..., 1]
cos = freqs_cos[pos, :half]
sin = freqs_sin[pos, :half]

out_even = even * cos - odd * sin
out_odd = even * sin + odd * cos
out = stack(out_even, out_odd).flatten(-2)
```

断言与 complex reference 一致。

### 3. Inverse RoPE apply 与 `inverse=True` 对齐

同样构造随机 attention output rope slice，比较：

```python
model.apply_rotary_emb(x_ref, freqs_cis, inverse=True)
```

和 real cos/sin inverse 展开公式：

```text
even' = even * cos + odd * sin
odd'  = odd * cos - even * sin
```

用于覆盖 `decode_sparse_attn.py` / `prefill_sparse_attn.py` 的 de-rotation 语义。

### 4. Position materialization 测试

覆盖：

- full table -> token-local `[T, ROPE_HEAD_DIM]`
- full table -> per-batch half `[B, ROPE_HEAD_DIM // 2]`
- compressed write position：

```text
cmp_pos = write_pos + 1 - COMPRESS_RATIO
```

断言 materialized rows 与 full table 对应行一致。

### 5. Regression tests for existing kernels

替换 fixture 初始化后，至少跑以下路径各一个 standalone kernel：

- SWA：`decode_attention_swa.py` 或 `prefill_attention_swa.py`
- CSA：`decode_attention_csa.py` 或 `prefill_attention_csa.py`
- HCA：`decode_attention_hca.py` 或 `prefill_attention_hca.py`
- token-local sparse：`decode_sparse_attn.py` 或 `prefill_sparse_attn.py`
- standalone qkv：`qkv_proj_rope.py`
- compressor/indexer fixture：
  - `prefill_compressor_ratio4.py`
  - `prefill_compressor_ratio128.py`
  - `prefill_indexer_compressor.py`
  - decode compressor/indexer standalone tests after confirming correct position selection

## 验收标准

- V4 path 有统一的 host-side RoPE/YaRN table builder。
- 不再依赖 synthetic `cos(arange * 1e-3)` 来模拟 DeepSeek RoPE。
- 生成的 `freqs_cos/freqs_sin` 与 `model.py::precompute_freqs_cis()` 的 real/imag 部分一致。
- V4 正向 RoPE 和 inverse RoPE 展开公式与 `model.py::apply_rotary_emb()` / `inverse=True` 对齐。
- SWA 和 compressed layers 使用正确且不同的 RoPE profiles。
- kernel signatures 尽量保持不变。
- 不新增用于整表生成的 PyPTO device kernel。
